Issues 2038

vniehues opened the pull request:
### Description of Change
Support for Navigation into sub-pages of TabbedPage.CurrentPage using selectedTab
```
TabbedPage?selectedTab=Tab2|ViewA|ViewB|ViewC
```

### Bugs Fixed
#2038

### API Changes
The signature of multiple internal methods of PageNavigationService were changed to async Task and call sites were awaited.

### Behavioral Changes
PageNavigationService will now parse the selectedTab portion of the string/Uri and split it based on the pipe. These changes were concentrated within the TabbedPageSelectTab method.

### PR Checklist

- [x] Has tests (if omitted, state reason in description)
- [X] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/PrismLibrary/Prism?pullRequest=2201) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/PrismLibrary/Prism?pullRequest=2201) before we can accept your contribution.<br/><br/>:x: vniehues [sign now](https://cla.dotnetfoundation.org/PrismLibrary/Prism?pullRequest=2201)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/PrismLibrary/Prism?pullRequest=2201) it.</sub>

dansiegel commented:
It appears that this is failing due to some issues with the hosted agent not currently defaulting to Xcode 12... this will affect all builds until the agents update. While there are some options to update default I'll need to think about how I want to handle that moving forward.

dansiegel commented on src/Forms/Prism.Forms/Navigation/NavigationException.cs:
```diff
@@ -12,6 +12,7 @@ public class NavigationException : Exception
         public const string IConfirmNavigationReturnedFalse = "IConfirmNavigation returned false";
         public const string NoPageIsRegistered = "No Page has been registered with the provided key";
         public const string ErrorCreatingPage = "An error occurred while resolving the page. This is most likely the result of invalid XAML or other type initialization exception";
+        public const string AmbiguousSelectedTab = "The requested selected tab matches more than one child page";
```
It appears that the changes to the NavigationException don't seem to have been implemented anywhere... we'll need to locate the appropriate place in the NavigationService to add the new AmbiguousPathException as the inner exception for this use case

vniehues replied:
I don't think that this is needed since Prism will just navigate to the first matching Tab and continue navigating within that tab.

In [your update](https://brianlagunas.com/whats-new-in-prism-for-xamarin-forms-7-0/) for Prism 7.0 you wrote, that having multiple NavigationPages is not an issue anymore since we are using the selectedTab parameter.

dansiegel commented on src/Forms/Prism.Forms/Navigation/NavigationException.cs:
```diff
@@ -29,4 +30,14 @@ public NavigationException(string message, Page page, Exception innerException)
 
         public Page Page { get; }
     }
+
+    public class AmbiguousPathException : Exception
+    {
+        public string Name { get; }
+
+        public AmbiguousPathException(string name)
+        {
+            Name = name;
+        }
+    }
```
this should be in it's own file... see notes about using it as the inner exception to the NavigationException

vniehues replied:
I don't think that this is needed since Prism will just navigate to the first matching Tab and continue navigating within that tab.

In [your update](https://brianlagunas.com/whats-new-in-prism-for-xamarin-forms-7-0/) for Prism 7.0 you wrote, that having multiple NavigationPages is not an issue anymore since we are using the selectedTab parameter.

dansiegel commented on src/Forms/Prism.Forms/Navigation/PageNavigationService.cs:
```diff
@@ -852,9 +852,9 @@ protected virtual Page CreatePageFromSegment(string segment)
-            PageUtilities.SetAutowireViewModel(page);
-            _pageBehaviorFactory.ApplyPageBehaviors(page);
-            ConfigurePages(page, segment);
+                PageUtilities.SetAutowireViewModel(page);
+                _pageBehaviorFactory.ApplyPageBehaviors(page);
+                await ConfigurePages(page, segment);
```
not sure what's up with the spacing here, but this should be reset to be at the right indent level

vniehues replied:
fixed

dansiegel commented on src/Forms/Prism.Forms/Navigation/NavigationException.cs:
```diff
@@ -29,4 +30,14 @@ public NavigationException(string message, Page page, Exception innerException)
 
         public Page Page { get; }
     }
+
+    public class AmbiguousPathException : Exception
+    {
+        public string Name { get; }
+
+        public AmbiguousPathException(string name)
+        {
+            Name = name;
+        }
+    }
```
this should be moved to it's own file and used as the inner exception to a NavigationException for the appropriate use case

dansiegel commented on src/Forms/Prism.Forms/Navigation/NavigationException.cs:
```diff
@@ -12,6 +12,7 @@ public class NavigationException : Exception
         public const string IConfirmNavigationReturnedFalse = "IConfirmNavigation returned false";
         public const string NoPageIsRegistered = "No Page has been registered with the provided key";
         public const string ErrorCreatingPage = "An error occurred while resolving the page. This is most likely the result of invalid XAML or other type initialization exception";
+        public const string AmbiguousSelectedTab = "The requested selected tab matches more than one child page";
```
this case doesn't appear to have actually been handled. please be sure to update the appropriate location in the NavigationService

vniehues commented:
> It appears that this is failing due to some issues with the hosted agent not currently defaulting to Xcode 12... this will affect all builds until the agents update. While there are some options to update default I'll need to think about how I want to handle that moving forward.

Yeah that's what i figured. Thank you for confirming it. 

I'll have a look at your change requests and report back afterwards

brianlagunas commented:
The API changes section of the PR description is incorrect. That is the current navigation API, not the proposed API.

vniehues commented:
@brianlagunas I fixed that by basically copying the original PRs description because thats still correct.
I also rebased it on your current master again.

dansiegel commented:
@vniehues this PR is going to be on hold until we release Prism 8. This represents a significant change that hasn't been available in a preview and it would be reckless of us to merge it as we're prepping for a stable release.

dansiegel commented:
@vniehues now that Prism 8 is out we'll be able to give this some attention. If you can please do a rebase and resolve the conflicts.

vniehues commented:
@dansiegel Sure thing!
I rebased this PR on your current master, fixed all conflicts, made sure there were no build errors and ran the ui-tests (all passed).
Some of the e2e tests which were failing earlier are now passing as well. 
Do you think the failed tests are due to a build-configuration error again?

dansiegel commented:
/azp run

dansiegel commented:
/azp run

dansiegel commented on src/Forms/Prism.Forms/Navigation/PageNavigationService.cs:
```diff
@@ -952,41 +953,56 @@ void ConfigureCarouselPage(CarouselPage carouselPage, string segment)
             {
-                var selectedTabType = PageNavigationRegistry.GetPageType(UriParsingHelper.GetSegmentName(selectedTab));
+                string segment;
+                Queue<string> tabSegements = null;
+                if (selectedTab.Contains('|'))
+                {
+                    tabSegements = new Queue<string>(selectedTab.Split('|'));
+                    segment = tabSegements.Dequeue();
+                }
+                else
+                {
+                    segment = selectedTab;
+                }
+
+                var selectedTabType =
+                    PageNavigationRegistry.GetPageType(UriParsingHelper.GetSegmentName(segment));
```
this currently isn't iterating like it needs to so if you have `NavigationPage|ViewA|ViewB|ViewC` you only end up with `NavigationPage|ViewA`
